possible bug in out_location

I have a public project https://github.com/tcncloud/api and I need to use it as dependency (I have setup a test in https://github.com/tcncloud/please-import-paths). 
Everything compiles and checks out in the api project, but when I reference it into my down stream project I get this error:
```
$ plz build ///api//v1beta:_v1beta#protoc                                                                                                                                                     09:54:42
Build stopped after 4s. 1 target failed:
    ///api//v1beta:_v1beta#protoc
Error building target ///api//v1beta:_v1beta#protoc: exit status 1
internal/modules/cjs/loader.js:638
    throw err;
    ^

Error: Cannot find module '/home/tom/tmp/please-import-paths/plz-out/tmp/api/v1beta/_v1beta#protoc._build/plz-out/gen/api/third_party/js/protoc-gen-ts'
    at Function.Module._resolveFilename (internal/modules/cjs/loader.js:636:15)
    at Function.Module._load (internal/modules/cjs/loader.js:562:25)
    at Function.Module.runMain (internal/modules/cjs/loader.js:831:12)
    at startup (internal/bootstrap/node.js:283:19)
    at bootstrapNodeJSCore (internal/bootstrap/node.js:622:3)
--ts_out: protoc-gen-ts: Plugin failed with status code 1.

```

Tracing down the error everything comes down to how the out_location is treated for the same target in both locations 
In the api project:
```
$ plz build --shell //third_party/js:protoc-gen-ts                                                                                                                                                        09:56:23
Temp directories prepared, total time 40ms:
  //third_party/js:protoc-gen-ts: plz-out/tmp/third_party/js/protoc-gen-ts._build
    Command: echo "#!/usr/bin/env bash" >> $OUT && echo "${TOOLS_NODE} /home/tom/dev/tcn/m/api/plz-out/gen/third_party/js/protoc-gen-ts/$(${TOOLS_JQ} -r '.bin["protoc-gen-ts"]' /home/tom/dev/tcn/m/api/plz-out/gen/third_party/js/protoc-gen-ts/package.json) $@" >> $OUT

```
the path rendered out here is `/home/tom/dev/tcn/m/api/plz-out/gen/third_party/js/protoc-gen-ts/`
In the downstream project looks like this: 
```
$ plz build --shell ///api//third_party/js:protoc-gen-ts                                                                                                                                      09:58:41
Temp directories prepared, total time 50ms:
  ///api//third_party/js:protoc-gen-ts: plz-out/tmp/api/third_party/js/protoc-gen-ts._build
    Command: echo "#!/usr/bin/env bash" >> $OUT && echo "${TOOLS_NODE} plz-out/gen/api/third_party/js/protoc-gen-ts/$(${TOOLS_JQ} -r '.bin["protoc-gen-ts"]' plz-out/gen/api/third_party/js/protoc-gen-ts/package.json) $@" >> $OUT

```
where the path rendered looks like this: `plz-out/gen/api/third_party/js/protoc-gen-ts/`

I try to replicate this by adding a test into `src/core/command_replacements_test.go` but so far I wasn't able to replicate it.

```
func TestOutLocation(t *testing.T) {
	target2 := makeTarget("//path/to:target2", "", nil)
	target1 := makeTarget("//path/to:target1", "ln -s $(out_location //path/to:target2) ${OUT}", target2)

	expected := "ln -s plz-out/gen/path/to/target2.py ${OUT}"
	assert.Equal(t, expected, replaceSequences(state, target1))
}

func TestOutLocationCrossProject(t *testing.T) {
	target2 := makeTarget("///project1//path/to:target2", "", nil)
	target1 := makeTarget("//path/to:target1", "ln -s $(out_location ///project1//path/to:target2) ${OUT}", target2)

	expected := "ln -s plz-out/gen/project1/path/to/target2.py ${OUT}"
	assert.Equal(t, expected, replaceSequences(state, target1))
}

```

--- a/src/core/build_label.go
+++ b/src/core/build_label.go
@@ -122,7 +122,7 @@ func validateTargetName(name string) bool {
 
 // ParseBuildLabel parses a single build label from a string. Panics on failure.
 func ParseBuildLabel(target, currentPath string) BuildLabel {
-	label, err := TryParseBuildLabel(target, currentPath)
+	label, err := TryParseBuildLabel(target, currentPath, "")
 	if err != nil {
 		panic(err)
 	}
@@ -130,8 +130,8 @@ func ParseBuildLabel(target, currentPath string) BuildLabel {
 }
 
 // TryParseBuildLabel attempts to parse a single build label from a string. Returns an error if unsuccessful.
-func TryParseBuildLabel(target, currentPath string) (BuildLabel, error) {
-	if pkg, name, subrepo := parseBuildLabelParts(target, currentPath, nil); name != "" {
+func TryParseBuildLabel(target, currentPath, subrepo string) (BuildLabel, error) {
+	if pkg, name, subrepo := parseBuildLabelParts(target, currentPath, subrepo); name != "" {
 		return BuildLabel{PackageName: pkg, Name: name, Subrepo: subrepo}, nil
 	}
 	return BuildLabel{}, fmt.Errorf("Invalid build label: %s", target)
@@ -140,7 +140,7 @@ func TryParseBuildLabel(target, currentPath string) (BuildLabel, error) {
 // ParseBuildLabelContext parses a build label in the context of a package.
 // It panics on error.
 func ParseBuildLabelContext(target string, pkg *Package) BuildLabel {
-	if p, name, subrepo := parseBuildLabelParts(target, pkg.Name, pkg.Subrepo); name != "" {
+	if p, name, subrepo := parseBuildLabelParts(target, pkg.Name, pkg.SubrepoName); name != "" {
 		if subrepo == "" && pkg.Subrepo != nil && (target[0] != '@' && !strings.HasPrefix(target, "///")) {
 			subrepo = pkg.Subrepo.Name
 		} else {
@@ -154,7 +154,7 @@ func ParseBuildLabelContext(target string, pkg *Package) BuildLabel {
 
 // parseBuildLabelParts parses a build label into the package & name parts.
 // If valid, the name string will always be populated; the package string might not be if it's a local form.
-func parseBuildLabelParts(target, currentPath string, subrepo *Subrepo) (string, string, string) {
+func parseBuildLabelParts(target, currentPath, subrepo string) (string, string, string) {
 	if len(target) < 2 { // Always must start with // or : and must have at least one char following.
 		return "", "", ""
 	} else if target[0] == ':' {
@@ -164,10 +164,10 @@ func parseBuildLabelParts(target, currentPath string, subrepo *Subrepo) (string,
 		return currentPath, target[1:], ""
 	} else if target[0] == '@' {
 		// @subrepo//pkg:target or @subrepo:target syntax
-		return parseBuildLabelSubrepo(target[1:], currentPath, subrepo)
+		return parseBuildLabelSubrepo(target[1:], currentPath)
 	} else if strings.HasPrefix(target, "///") {
 		// ///subrepo/pkg:target syntax.
-		return parseBuildLabelSubrepo(target[3:], currentPath, subrepo)
+		return parseBuildLabelSubrepo(target[3:], currentPath)
 	} else if target[0] != '/' || target[1] != '/' {
 		return "", "", ""
 	} else if idx := strings.IndexRune(target, ':'); idx != -1 {
@@ -177,7 +177,7 @@ func parseBuildLabelParts(target, currentPath string, subrepo *Subrepo) (string,
 		if !validatePackageName(pkg) || !validateTargetName(name) || name == "..." {
 			return "", "", ""
 		}
-		return pkg, name, ""
+		return pkg, name, subrepo
 	} else if !validatePackageName(target[2:]) {
 		return "", "", ""
 	}
@@ -185,13 +185,13 @@ func parseBuildLabelParts(target, currentPath string, subrepo *Subrepo) (string,
 	if strings.HasSuffix(target, "/...") {
 		return strings.TrimRight(target[2:len(target)-3], "/"), "...", ""
 	} else if idx := strings.LastIndexByte(target, '/'); idx != -1 {
-		return target[2:], target[idx+1:], ""
+		return target[2:], target[idx+1:], subrepo
 	}
-	return target[2:], target[2:], ""
+	return target[2:], target[2:], subrepo
 }
 
 // parseBuildLabelSubrepo parses a build label that began with a subrepo symbol (either @ or ///).
-func parseBuildLabelSubrepo(target, currentPath string, subrepo *Subrepo) (string, string, string) {
+func parseBuildLabelSubrepo(target, currentPath string) (string, string, string) {
 	idx := strings.Index(target, "//")
 	if idx == -1 {
 		// if subrepo and target are the same name, then @subrepo syntax will also suffice
@@ -202,7 +202,7 @@ func parseBuildLabelSubrepo(target, currentPath string, subrepo *Subrepo) (strin
 			return "", target, target
 		}
 	}
-	pkg, name, _ := parseBuildLabelParts(target[idx:], currentPath, subrepo)
+	pkg, name, _ := parseBuildLabelParts(target[idx:], currentPath, "")
 	return pkg, name, target[:idx]
 }
 
@@ -212,7 +212,7 @@ func parseMaybeRelativeBuildLabel(target, subdir string) (BuildLabel, error) {
 	// Try the ones that don't need locating the repo root first.
 	startsWithColon := strings.HasPrefix(target, ":")
 	if !startsWithColon {
-		if label, err := TryParseBuildLabel(target, ""); err == nil || strings.HasPrefix(target, "//") {
+		if label, err := TryParseBuildLabel(target, "", ""); err == nil || strings.HasPrefix(target, "//") {
 			return label, err
 		}
 	}
@@ -223,10 +223,10 @@ func parseMaybeRelativeBuildLabel(target, subdir string) (BuildLabel, error) {
 		subdir = initialPackage
 	}
 	if startsWithColon {
-		return TryParseBuildLabel(target, subdir)
+		return TryParseBuildLabel(target, subdir, "")
 	}
 	// Presumably it's just underneath this directory (note that if it was absolute we returned above)
-	return TryParseBuildLabel("//"+path.Join(subdir, target), "")
+	return TryParseBuildLabel("//"+path.Join(subdir, target), "", "")
 }
 
 // ParseBuildLabels parses a bunch of build labels from strings. It dies on failure.
@@ -334,7 +334,7 @@ func (label *BuildLabel) UnmarshalFlag(value string) error {
 // UnmarshalText implements the encoding.TextUnmarshaler interface.
 // This is used by gcfg to unmarshal the config files.
 func (label *BuildLabel) UnmarshalText(text []byte) error {
-	l, err := TryParseBuildLabel(string(text), "")
+	l, err := TryParseBuildLabel(string(text), "", "")
 	*label = l
 	return err
 }

--- a/src/core/build_label_test.go
+++ b/src/core/build_label_test.go
@@ -120,8 +120,8 @@ func TestSubrepoLabel(t *testing.T) {
 func TestParseBuildLabelParts(t *testing.T) {
 	target1 := "///unittest_cpp//:unittest_cpp"
 	targetNewSyntax := "@unittest_cpp"
-	pkg, name, subrepo := parseBuildLabelParts(target1, "/", nil)
-	pkg2, name2, subrepo2 := parseBuildLabelParts(targetNewSyntax, "/", nil)
+	pkg, name, subrepo := parseBuildLabelParts(target1, "/", "")
+	pkg2, name2, subrepo2 := parseBuildLabelParts(targetNewSyntax, "/", "")
 	assert.Equal(t, pkg, "")
 	assert.Equal(t, pkg2, "")
 	assert.Equal(t, name, "unittest_cpp")

--- a/src/core/command_replacements.go
+++ b/src/core/command_replacements.go
@@ -155,7 +155,10 @@ func replaceSequencesInternal(state *BuildState, target *BuildTarget, command st
 // replaceSequence replaces a single escape sequence in a command.
 func replaceSequence(state *BuildState, target *BuildTarget, in string, runnable, multiple, dir, outPrefix, hash, test bool) string {
 	if LooksLikeABuildLabel(in) {
-		label := ParseBuildLabel(in, target.Label.PackageName)
+		label, err := TryParseBuildLabel(in, target.Label.PackageName, target.Label.Subrepo)
+		if err != nil {
+			panic(err)
+		}
 		return replaceSequenceLabel(state, target, label, in, runnable, multiple, dir, outPrefix, hash, test, true)
 	}
 	for _, src := range sourcesOrTools(target, runnable) {

--- a/src/core/command_replacements_test.go
+++ b/src/core/command_replacements_test.go
@@ -116,6 +116,17 @@ func TestToolReplacement(t *testing.T) {
 	assert.Equal(t, expected, cmd)
 }
 
+func TestToolReplacementSubrepo(t *testing.T) {
+	target2 := makeTarget("///subrepo//path/to:target2", "blah", nil)
+	target1 := makeTarget("///subrepo//path/to:target1", "$(location //path/to:target2)", target2)
+	target1.Tools = append(target1.Tools, target2.Label)
+
+	wd, _ := os.Getwd()
+	expected := quote(path.Join(wd, "plz-out/gen/path/to/target2.py"))
+	cmd, _ := ReplaceSequences(state, target1, target1.Command)
+	assert.Equal(t, expected, cmd)
+}
+
 func TestDirReplacement(t *testing.T) {
 	target2 := makeTarget("//path/to:target2", "blah", nil)
 	target2.AddOutput("blah2.txt")

--- a/src/core/command_replacements_test.go
+++ b/src/core/command_replacements_test.go
@@ -122,7 +122,7 @@ func TestToolReplacementSubrepo(t *testing.T) {
 	target1.Tools = append(target1.Tools, target2.Label)
 
 	wd, _ := os.Getwd()
-	expected := quote(path.Join(wd, "plz-out/gen/path/to/target2.py"))
+	expected := quote(path.Join(wd, "plz-out/gen/subrepo/path/to/target2.py"))
 	cmd, _ := ReplaceSequences(state, target1, target1.Command)
 	assert.Equal(t, expected, cmd)
 }

--- a/src/core/package.go
+++ b/src/core/package.go
@@ -118,7 +118,7 @@ func (pkg *Package) HasSubinclude(label BuildLabel) bool {
 
 // SubrepoArchName returns a subrepo name, modified for the architecture of this package if it's not the host.
 func (pkg *Package) SubrepoArchName(subrepo string) string {
-	if subrepo != "" && pkg.Subrepo != nil && pkg.Subrepo.IsCrossCompile {
+	if subrepo != "" && pkg.Subrepo != nil && pkg.Subrepo.IsCrossCompile && pkg.SubrepoName != subrepo {
 		return subrepo + "_" + pkg.SubrepoName
 	}
 	return subrepo

--- a/src/parse/parse_step.go
+++ b/src/parse/parse_step.go
@@ -91,7 +91,7 @@ func checkSubrepo(tid int, state *core.BuildState, label, dependent core.BuildLa
 	if handled, err := parseSubrepoPackage(tid, state, sl.PackageName, dependent.Subrepo, label); handled && err == nil {
 		return state.Graph.Subrepo(label.Subrepo), nil
 	}
-	return nil, fmt.Errorf("Subrepo %s is not defined", label.Subrepo)
+	return nil, fmt.Errorf("Subrepo %s is not defined (referenced by %s)", label.Subrepo, dependent)
 }
 
 // parseSubrepoPackage parses a package to make sure subrepos are available.

--- a/tools/build_langserver/lsp/completion.go
+++ b/tools/build_langserver/lsp/completion.go
@@ -29,7 +29,7 @@ func (h *Handler) completeLabel(doc *doc, partial string, line, col int) (*lsp.C
 		list := &lsp.CompletionList{}
 		pkgName := path.Base(doc.Filename)
 		pkgLabel := core.BuildLabel{PackageName: pkgName, Name: "all"}
-		label, err := core.TryParseBuildLabel(labelName, pkgName)
+		label, err := core.TryParseBuildLabel(labelName, pkgName, pkgLabel.Subrepo)
 		if err != nil {
 			return nil, err
 		}

--- a/tools/build_langserver/lsp/diagnostics.go
+++ b/tools/build_langserver/lsp/diagnostics.go
@@ -35,7 +35,7 @@ func (h *Handler) diagnostics(d *doc, ast []*asp.Statement) []lsp.Diagnostic {
 	asp.WalkAST(ast, func(expr *asp.Expression) bool {
 		if expr.Val != nil && expr.Val.String != "" {
 			if s := stringLiteral(expr.Val.String); core.LooksLikeABuildLabel(s) {
-				if l, err := core.TryParseBuildLabel(s, pkgLabel.PackageName); err == nil {
+				if l, err := core.TryParseBuildLabel(s, pkgLabel.PackageName, pkgLabel.Subrepo); err == nil {
 					if l.IsAllTargets() || l.IsAllSubpackages() {
 						// Can't emit any useful info for these.
 						// TODO(peterebden): If we know what argument we were in we could emit info

